torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -249,22 +249,12 @@ def _shard_parameters(self) -> None:
                 self.numel_padded_per_param.append(0)
                 continue
 
-            # Save the original storage and free it later on.
-            # Since we're modifying the tensor's storage directly,
-            # make sure the tensor is the sole occupant of the storage.
-            assert (
-                p.storage_offset() == 0
-            ), "The tensor is not the sole occupant of the storage."
-            orig_storage = p.storage()
-
-            # Replace p with the relevant shard.
-            local_shard, num_padded = self._get_shard(p)
-            p.set_(local_shard)  # type: ignore[call-overload]
```

Comment:
> 'A view was created in no_grad mode and its base or another view of its base has been modified inplace with grad mode enabled. Given that this use case is ambiguous and error-prone, it is forbidden. You can clarify your code by moving both the view and the inplace either both inside the no_grad block (if you don't want the inplace to be tracked) or both outside (if you want the inplace to be tracked).'

Do we know where was the inplace modification under grad mode? And would I be correct if I assume that we only hit this error with the new tests?

Reply:
yes, only hit this error for multiprocessing tests, I'm not 100% clear about how the pickle, data transfer works inside torch.multiprocessing. basically if using p.set_(), autograd will do more rigid checks, while p.data will not be able to be checked.
I changed p.data implementation (FairScale implementation) to p.set_() in our first PR, now changing it back for passing torch.multiprocessing tests

Reply:
Maybe @albanD might know what subtle differences can exist between `p.data` and `p.set` that might result in the error @mrshenli posted above for multiprocessing tests?

Reply:
`p.data` is hiding everything from autograd so no error checking happens and anything can be wrong.
`p.set_()` on the other hand is properly registered as an inplace update onto p.